File: java/src/main/java/problems1_100/RemoveDuplicatesFromSortedList2.java
package problems1_100;

import concepts.SinglyListNode;

enum RemoveDuplicatesFromSortedList2 {
    DEFAULT {
        @Override
        SinglyListNode deleteDuplicates(SinglyListNode head) {
            SinglyListNode root = new SinglyListNode(-101, head);
            SinglyListNode previous = root;
            SinglyListNode current = head;
            while (current != null) {
                if (current.next != null && current.next.val == current.val) {
                    while (current.next != null && current.next.val == current.val) {
                        current = current.next;
                    }
                    previous.next = current.next;
                } else {
                    previous = previous.next;
                }
                current = current.next;
            }
            return root.next;
        }
    };

    abstract SinglyListNode deleteDuplicates(SinglyListNode head);
}

File: kotlin/src/main/kotlin/problems1_100/RemoveDuplicatesFromSortedList2.kt
package problems1_100

import concepts.SinglyListNode

enum class RemoveDuplicatesFromSortedList2 {
    DEFAULT {
        override fun deleteDuplicates(head: SinglyListNode?): SinglyListNode? {
            val root = SinglyListNode(-101, head)
            var previous: SinglyListNode? = root
            var current = head
            while (current != null) {
                if (current.next?.`val` == current.`val`) {
                    while (current!!.next?.`val` == current.`val`) {
                        current = current.next
                    }
                    previous!!.next = current.next
                } else {
                    previous = previous!!.next
                }
                current = current.next
            }
            return root.next
        }
    },
    ;

    abstract fun deleteDuplicates(head: SinglyListNode?): SinglyListNode?
}
